### PR TITLE
Utility functions to read/write from/to Dataiku objects

### PR DESCRIPTION
**Pull request by alexcombessie**

### CODE REVIEW

**Comment by alexcombessie** [comment]
FYI @JaneBellaiche, the unit tests are broken here because of an unrelated change brought by @HenriChabert (acdf1c430cc702c6387dc9ac5862a4cbaac7fd71). I have let him know about it already. 

There are no unit tests for the dku_io_utils module, because it depends on the internal Dataiku API (which cannot be installed with pip).

**Comment by JaneBellaiche** [comment]
[ch-65991]

File: dkulib/dku_io_utils/chunked_processing.py
Comment (Muennighoff) [inline dkulib/dku_io_utils/chunked_processing.py:51]: Perhaps a stupid question - but what's the key differentiation from DataFrameParallelizer in [dkulib/parallelizer](https://github.com/dataiku/dss-plugin-dkulib/blob/feature/parallelizer/dkulib/parallelizer/parallelizer.py)?

I see:
- One chunk in memory at a time - Might also be possible to do in DataFrameParallelizer, no?
- Parallelization -- can be off if workers is just 1
- Feeding in a dataiku.Dataset vs pd.DataFrame - could be changed in one or the other

Wouldn't it be easier for users to have just one go-to wrapper for applying functions to dataiku datasets?

File: dkulib/dku_io_utils/column_descriptions.py
Comment (Muennighoff) [inline dkulib/dku_io_utils/column_descriptions.py:34]: I think it would be more intuitive to first set all column descriptions to those of the input & then update the ones which have changed in the output rather than the other way round

In the current implementation if I feed in an output dataset with a new description that was empty previously and I also feed in an input dataset because I want to keep some descriptions, it replaces all my new descriptions with "" (Perhaps I'm doing something wrong)

Comment (alexcombessie) [reply dkulib/dku_io_utils/column_descriptions.py:34]: Interesting, that seems unexpected. Can you detail what you had as input and what you received as output?

File: dkulib/dku_io_utils/chunked_processing.py
Comment (alexcombessie) [reply dkulib/dku_io_utils/chunked_processing.py:51]: Good remark! While there are some similarities, there are actually several differences:
1. DataFrameParallelizer works on pandas.DataFrame, while process_dataset_chunks works on dataiku.Dataset. 
2. DataFrameParallelizer can handle errors, so is well suited for failure-prone API calls. In contrast, process_dataset_chunks will fail upon any error from the underlying function.
3. DataFrameParallelizer enforces a set output schema, designed for API calls with responses and errors. Meanwhile, process_dataset_chunks lets the user set the schema of the output DataFrame.
   - For the record, this is due to the fact both approaches were designed for different needs: DataFrameParallelizer was designed for API-related plugins, whereas process_dataset_chunks was designed for pure Python NLP processing.
4. DataFrameParallelizer can parallelize (row-by-row or by batches of rows) whereas process_dataset_chunks is working sequentially

I chose to write both modules separately mainly because:
(i) I find value in the separation of concerns between internal Dataiku API stuff and pure Python pandas stuff. In particular, we don't currently have a way to test dkulib modules if they rely on the internal Dataiku API.
(ii) End users can freely choose to use one or the other, or both at the same time:
```
process_dataset_chunks(
    input_dataset=dataiku.Dataset("input"),
    output_dataset=dataiku.Dataset("output"),
    func=parallelizer.run
)
```

Now, I agree there would be long-term value in merging the two concepts in a single module. I would definitely log that as an idea for future improvement. Not necessarily a must-have for right now (if it's not broken, ...) but definitely something to keep in mind.

What do you think?

File: dkulib/dku_io_utils/column_descriptions.py
Comment (Muennighoff) [reply dkulib/dku_io_utils/column_descriptions.py:34]: Sure here's the input:

Output dataset; See schema:
[{'name': 'order_date', 'type': 'string'}, {'name': 'pages_visited', 'type': 'bigint'}, {'name': 'order_id', 'type': 'string'}, {'name': 'customer_id', 'type': 'string'}, {'name': 'tshirt_category', 'type': 'string'}, {'name': 'tshirt_price', 'type': 'double'}, {'name': 'tshirt_quantity', 'type': 'bigint'}] 

Column descriptions:
{'order_date': 'Foo'} 

Input dataset; See schema:
[{'name': 'order_date', 'type': 'string', 'comment': ''}, {'name': 'pages_visited', 'type': 'string'}, {'name': 'order_id', 'type': 'string'}, {'name': 'customer_id', 'type': 'string'}, {'name': 'tshirt_category', 'type': 'string'}, {'name': 'tshirt_price', 'type': 'string'}, {'name': 'tshirt_quantity', 'type': 'string'}]

Because the comment is an empty string for order_date it doesnt add the new description. The empty string results from me applying it a couple times I guess

Comment (alexcombessie) [reply dkulib/dku_io_utils/column_descriptions.py:34]: Very good catch! Let me rewrite the function as you advised. Should be simpler and fix the case you described.

File: dkulib/dku_io_utils/chunked_processing.py
Comment (Muennighoff) [reply dkulib/dku_io_utils/chunked_processing.py:51]: Good with me!
Thanks for the detailed explanation 😊

File: dkulib/dku_io_utils/column_descriptions.py
Comment (alexcombessie) [reply dkulib/dku_io_utils/column_descriptions.py:34]: Hi Niklas, I am not able to reproduce the issue. Do you have an example project export or a screen recording to explain what you mean? Thanks!

Comment (alexcombessie) [reply dkulib/dku_io_utils/column_descriptions.py]: Thanks a lot Niklas. I have been able to reproduce thanks to the project export you shared on Slack. Here is a refactored version according to your feedback: 46c9d02
Let me know what you think.

**Comment by alexcombessie** [comment]
Hi @Muennighoff,

I re-requested a review since I refactored the column descriptions module.

FYI I had to merge the parallelizer branch in order to fix the failing unit tests.

**Review by Muennighoff — APPROVED** [review]
LGTM & all funcs working on my end👍

**Comment by alexcombessie** [comment]
Merci Niklas! I will merge once the parallelizer is OK and merged

**Pull request merged**
